p2p/src/codec.rs

```diff
@@ -0,0 +1,268 @@
+use crate::cybershake::PublicKey;
+use crate::peer::PeerAddr;
+use crate::{PeerID, PeerMessage};
+use bytes::{Buf, BufMut, BytesMut};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use std::io;
+use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6};
+use tokio_util::codec::{Decoder, Encoder};
+
+pub struct MessageEncoder;
+
+impl Encoder<PeerMessage> for MessageEncoder {
+    type Error = io::Error;
+
+    fn encode(&mut self, item: PeerMessage, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        match item {
+            PeerMessage::Hello(u) => {
+                dst.put_u8(0); // Message type
+                const HELLO_MESSAGE_LEN: u64 = 2;
+                dst.put_u64(HELLO_MESSAGE_LEN);
+                dst.put_u16(u);
+            }
+            PeerMessage::Peers(p) => {
+                dst.put_u8(1); // Message type
+                let body = p.into_iter().fold(BytesMut::new(), |mut bytes, peer| {
```

Comment:
can we fold using the current `dst`, so we don't allocate another buffer? Also the code might be more straightforward: first write the length prefix, then each peer address.

Reply:
Changed: now we reserve 4 bytes for length prefix, than write body, then write length prefix at reserved place.